infrastructure/src/main/java/org/corfudb/infrastructure/NettyRequestRouter.java

```diff
@@ -0,0 +1,164 @@
+package org.corfudb.infrastructure;
+
+import com.google.common.collect.ImmutableList;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.io.IOUtils;
+import org.corfudb.protocols.API;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.MessageType;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Request;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Response;
+import org.corfudb.runtime.protocol.proto.CorfuProtocol.Header;
+import org.corfudb.runtime.view.Layout;
+
+import java.io.IOException;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+@Slf4j
+@ChannelHandler.Sharable
+public class NettyRequestRouter extends ChannelInboundHandlerAdapter implements IRequestRouter {
```

Comment:
Do we actually need a separate class for this? Is there a reason why we did not just add the following two functions to `IClientRouter`?

```
    <T> CompletableFuture<T> sendRequestAndGetCompletable(CorfuProtocol.Request.Builder reqBuilder);
    void sendRequest(CorfuProtocol.Request.Builder reqBuilder);
```

Reply:
This is for the server. I wanted to do this initially, but we also require a new implementation of channelRead. This became tricky because I have two router instances in my pipeline, as otherwise, we would require it to process both types of messages. Maybe as a workaround, we can combine the functionality of NettyCorfuMsgEncoder and NettyCorfuMsgDecoder into the router, so that we can simply worry about sending and receiving a ByteBuf? I am not sure If this has any other implications... I'll take a look more carefully and see what can be done.